nixos/modules/services/web-apps/glance.nix

```
@@ -0,0 +1,135 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.glance;
+
+  inherit (lib)
+    mkEnableOption
+    mkPackageOption
+    mkOption
+    mkIf
+    getExe
+    types
+    ;
+
+  settingsFormat = pkgs.formats.yaml { };
+in
+{
+  options.services.glance = {
+    enable = mkEnableOption "glance";
+    package = mkPackageOption pkgs "glance" { };
+
+    host = lib.mkOption {
+      description = "Glance bind address";
+      default = "127.0.0.1";
+      type = types.str;
+    };
+
+    port = lib.mkOption {
+      description = "Glance port to listen on";
+      default = 8080;
+      type = types.port;
+    };
+
+    settings = mkOption {
+      type = settingsFormat.type;
```

Maybe we could place these options within `settings`? It would remove the need for the `recursiveUpdate` further down.

```suggestion
    settings = mkOption {
      type = lib.types.submodule {
			  freeformType = settingsFormat.type;
			  options = {
          host = lib.mkOption {
            description = "Glance bind address";
            default = "127.0.0.1";
            type = types.str;
          };

          port = lib.mkOption {
            description = "Glance port to listen on";
            default = 8080;
            type = types.port;
          };
        };
			};
```

---

I thought about doing that, but I really wanted to keep `host` and `port` out of the `glance` configuration file. IMO, it seems better doing so, but perhaps I am totally wrong.

---

By all means :) I guess I should also clarify that this is a completely non-blocking suggestion.

But I'm interested to hear your thoughts. What are the merits of keeping these outside `settings`? Is it better for discoverability, or for similarity with other modules?

The only downside I can see is a scenario where someone has an existing configuration and rewrites it to nix, only to put everything in settings and be surprised that the `host` and `port` is overwritten by the `default` values on the other options. I guess it's unlikely, since most people probably use https://search.nixos.org or something similar to discover options, but I know I have been bitten by similar overwriting mechanisms before.

---

OK cool.

This PR introduce a new service. That service has a configuration file and is bound to an address and a port.

For transparency, I definitely think that those 2 network settings should be clearly separated from the config file.

When thinking twice about your proposal, it actually could be a good idea because `port` and `host` are still separated... but I'm having trouble seeing how to make it work since those settings needs to be wrapped under `server.port` and `server.host` in the `glance` config file.

Have you tried your proposal, is it working?

---

I tried it out with this patch: https://gist.github.com/h7x4/caf28ce4eda95ef3b4aa3bd5f9349d3c

I also edited the default `pages` value (with example) to exist within the submodule, so that it doesn't disappear the moment you set anything within `services.glance.settings`.

---

Thank you, I merged it :) now I understand your point and I totally agree with it! TIL !